Specify Span ID creation with sampling. (#998)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

The SDK defines the interface [`Sampler`](#sampler) as well as a set of
[built-in samplers](#built-in-samplers) and associates a `Sampler` with each [`TracerProvider`].

> ### SDK Span creation

Reviewer: I doubt that all of our SDK implementations are aligned here so it would make sense to add a spec compliance matrix entry "Trace and Span IDs are created as defined in [this section]".

Author: Done in bfd652f.

When asked to create a Span, the SDK MUST act as if doing the following in order:

> 1. If there is a valid parent trace ID, use it. Otherwise generate a new one (
>    note: this must be done before calling `ShouldSample`, because it expects
>    a valid trace Id as input).

Reviewer: ```suggestion
1. If there is a valid parent trace ID, use it. Otherwise generate a new one
   (note: this must be done before calling `ShouldSample`, because it expects
   a valid trace ID as input).
```

2. Query the `Sampler`'s [`ShouldSample`](#shouldsample) method.

> 3. If the decision is `DROP`, use the parent span ID or all-zero if there is none.

Author: That means: If there is no parent *and* the decision is drop, we will have a span with the new trace ID but an invalid span ID. The SpanContext as a whole is thus invalid. We could also specify that the generated trace ID should be discarded in that case.

Author: Actually this may be worth re-considering: Generating trace+span ID for an unsampled root has implications if the span is injected. If we generate an invalid spancontext (with zero span ID), we will inject nothing as we have an invalid traceparent header as per W3C (cf. also #753). This causes downstream components to potentially create traces / run sampling logic that they would not if they got an incoming unsampled traceparent header.

Reviewer 2: > We could also specify that the generated trace ID should be discarded in that case.

I'd be fine with this.

> we will inject nothing as we have an invalid traceparent header as per W3C (cf. also #753). This causes downstream components to potentially create traces

Could you elaborate?

Author: @carlosalberto We never inject an invalid SpanContext, at least using W3C. If we would generate a new ID instead of using zero, we would inject a valid spancontext with unset sampled flag. This will likely cause the receiver to suppress the trace. On the other hand, if we use zero, the downstream component will not receive any tracestate and will likely start a new trace.

Now that I write, this, I realize that this even applies to local child spans: Normally if the root span is not sampled, local child spans would be unsampled too, but with this they would become root spans.

This is bad. I think we will have to create Span *and* Trace ID even for "dropped" root spans after all. It seems there is a dependency of #864 "Option to allow "default" IDs for unsampled traces" on #753 "https://github.com/open-telemetry/opentelemetry-specification/issues/753". We would have to have some sentinel SpanContext that is "valid, unsampled, but without any IDs". That's out of scope here of course.

Author: Updated in 4006816. Maybe this demonstrates how easy it is to mess up here and that it is thus important to have this specified clearly? 😃

>    Otherwise (if the decision is not `DROP`) a new span ID MUST be generated.

Author: That means: RECORD_ONLY spans get a new span ID, always. I don't know what the purpose of RECORD_ONLY is (should we remove it?), but having different spans with the same trace+span ID seems worse than breaking the trace in case any children of the RECORD_ONLY span become RECORD_AND_SAMPLE .

>    a new span ID MUST be generated.

Reviewer: The current state of this PR sounds like the right thing to do to me, but I'm not entirely sure of the consequences.
Fellow @open-telemetry/technical-committee, please take a look (especially at previous comments https://github.com/open-telemetry/opentelemetry-specification/pull/998#discussion_r503141158) as well. We might need to sort this out before GA.

Reviewer 3: Yes. This is tricky, but I like @Oberon00's proposal that DROP sampling decisions at the root of a trace should generate a new trace ID and new span ID without the sampled bit set. Am I understanding?

Author: > Am I understanding?

Yes, "no valid parent span ID" is equivalent to a root span.

   see description of [`ShouldSample`'s](#shouldsample) return value below
   for how to set `IsRecording` and `Sampled` on the Span,
   and the [table above](#recording-sampled-reaction-table) on whether
   to pass the `Span` to `SpanProcessor`s.
   A non-recording span MAY be implemented using the same mechanism as when a

>    `Span` is created with no API-implementation installed
>    (sometimes called a `NoOpSpan` or `DefaultSpan`).

Reviewer: ```suggestion
   `Span` is created with no API-implementation installed (i.e., a [Propagated Span](api.md#propagated-span-creation)).
```

The SDK defines the interface [`Sampler`](#sampler) as well as a set of
[built-in samplers](#built-in-samplers) and associates a `Sampler` with each [`TracerProvider`].

### SDK Span creation

When asked to create a Span, the SDK MUST act as if doing the following in order:

> 1. If there is a valid parent trace ID, use it. Otherwise generate a new one

Reviewer 3: nit: replace "new one" with "new trace ID"?

Author: ```suggestion
1. If there is a valid parent trace ID, use it. Otherwise generate a new trace ID
```

   (note: this must be done before calling `ShouldSample`, because it expects
   a valid trace ID as input).
2. Query the `Sampler`'s [`ShouldSample`](#shouldsample) method.

> 3. If the decision is `DROP` and there is a valid parent span ID, use it.

Reviewer 3: nit: I suggest more verbosity: "... and there is a valid parent span ID, use it as the new context's span ID".

Author: ```suggestion
3. If the decision is `DROP` and there is a valid parent span ID, reuse it as the new `Span`'s span ID.
```

The SDK defines the interface [`Sampler`](#sampler) as well as a set of
[built-in samplers](#built-in-samplers) and associates a `Sampler` with each [`TracerProvider`].

### SDK Span creation

When asked to create a Span, the SDK MUST act as if doing the following in order:

> 1. If there is a valid parent trace ID, use it. Otherwise generate a new trace ID

Reviewer 4: > If there is a valid parent trace ID, use it.

This is actually not compatible with how many cloud providers treat external trace ID coming from a customer - providers may or may not trust such external IDs. So MUST is too strong here, there is an option for configurable behavior.

Author: I think MUST is exactly right here. We currently MUST reuse the parent trace ID, or otherwise not extract it at all at the extract level so that it does not even get to the Tracer/SDK. A configurable option to ignore a remote trace ID would be another concern and maybe a good argument for fusing the configurable ID generation into Sampler (#896)

Reviewer 4: I would disagree. Extracting and sampling are well-encapsulated functionalities, which are orthogonal to the decision on reusing or regenerating a trace ID. Trying to put this aspect into extraction/sampling is an abstraction leak. There are already multiple implementations of both extractors and samplers, they would all have to deal with this extra concern, consistently. I think it belongs to the Tracer, and be configurable.

Author: So basically, if SpanContext.IsRemote, drop the SpanContext? Or add an additional ParentFilter component that is queried before the Sampler and before deciding whether to generate a new ID?
I think this is out of scope here though. If such a component is added, we would add an additional step before the first like

> 1. Check if the parent span is valid for use by querying the configured ParentFilter.ShouldAccept with the parent Context as input. If the decision is REJECT, proceed as if no valid parent trace and span ID had been passed. Otherwise continue with the passed-in span and trace ID.

But this is really a separate issue. Right now if one wants to implement this the propagator is usable for that (even though you might argue it is not responsible). Other points to consider:
1. You cannot reject every IsRemote parent, that would make your distributed tracing non-distributed. You probably need some cryptographic signatures in place.
2. If, instead of cryptographic signatures, you want to rely on the origin of the span ID (e.g. source IP, etc) this would need to be done somewhere else where this information is still available, or the propagators would need to be adapted to enrich the context with such information
3. Do you want to forward an untrusted (Span)Context if you extract, then inject without creating a Span in-between? If not, then this check would need to be done on extract() (thus in a Propagator-(sub)component).

I think this is a separate, most-likely after-GA issue.

Reviewer 4: > So basically, if SpanContext.IsRemote, drop the SpanContext? Or add an additional ParentFilter component that is queried before the Sampler and before deciding whether to generate a new ID?

Not drop, I think cloud providers would capture it on the span, but they will start a new internal trace, whose ID may then be returned to the caller using (future) w3c response header.

> I think this is a separate, most-likely after-GA issue.

I don't mind this being addressed post-GA, my main concern was that the current MUST wording dictates that respecting the inbound trace ID is the only possible processing model. However, maybe that's OK for SDK v1.

Booked https://github.com/open-telemetry/opentelemetry-specification/issues/1188

   (note: this must be done before calling `ShouldSample`, because it expects
   a valid trace ID as input).

> 2. Query the `Sampler`'s [`ShouldSample`](#shouldsample) method.

Reviewer 4: This could be interpreted that a sampling decision MUST always be made. Perhaps worth clarifying that some samplers may retain/respect the propagated sampling decision instead of making a new one.

Author: Why do you think this could be interpreted this way? We do specify the ParentBased sampler below which will respect the parent's sampled flag in it's decision (translating sampled=true to RECORD and sampled=false to DROP is also a decision).

Reviewer 4: Yes, but you're using "forward reference" :-) When someone's reading this for the first time, and doesn't know about ParentBased sampler discussed later, they can easily interpret this as "make a sampling decision".

Author: I added a note in eb989b7.

> 3. If the decision is `DROP` and there is a valid parent span ID, reuse it as the new `Span`'s span ID.

Reviewer 5: Doesn't this mean two spans with the same id? I can imagine this could become extremely confusing to someone trying to debug things. As a user, I think I'd be super confused if I saw a new Span created with the same ID as the parent, even if that span might not end up going any where.

Author: It's two spans with the same ID the same way as calling extract and then inject creates a copy of the injected span (with the same ID). It is more of a hollow shell of a span than an actual span.

> even if that span might not end up going anywhere

This span cannot end up going anywhere. A dropped span is always non-recording, so there is nothing that could end up anywhere beside the span+trace ID itself (not even a parent span ID). What is more, SpanProcessors are not invoked with dropped spans.

RECORD_ONLY spans are a different story, and they do get a new ID, see https://github.com/open-telemetry/opentelemetry-specification/pull/998#discussion_r494451985

All that being said, I think I could agree that this may make the mental model of sampling a bit more complicated than just always creating a new ID. But I hope that the benefits, namely ability to resume sampling in a child and having the closes sampled span as parent, and a slight(?) performance improvement would be worth it.

Reviewer 5: Personally, I think I'd prefer attacking this problem via a smarter SpanContext that allowed propagation of the relevant bits, rather than generating weird internal spans for the sake of propagating the SpanContext.

Reviewer 5: Or, better yet, move the TraceState and TraceFlags out of the span altogether and propagate them separately in the Context, rather than hanging them off a span.

Author: What should startSpan return for a dropped span then?

Reviewer 5: Just the "invalid" span that it does today. all zeros for both of the ids.

Author: That would not work for the reasons outlined in https://github.com/open-telemetry/opentelemetry-specification/pull/998#discussion_r503141158, (the sampled-flag would be ignored). Also, Java always generates a new span ID currently (https://github.com/open-telemetry/opentelemetry-java/blob/34953d5feee5129b64c8cec04d23dc3be07a9be8/sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java#L204-L209), and the spec says nothing at all.

Reviewer 5: I agree with that. That's why I think we could solve this by moving the TraceState and flags out of the span, so samplers can use that information from the context, rather than having just the span to work with.

Author: Not sure if I follow. The TraceState and Flags are already in the SpanContext, not the Span. The problem is that we say that if trace OR span ID is zero, then we should disregard traceflags and tracestate. If we wanted to change that, the easiest and IMHO best way would be to simply say that only *the* invalid, empty SpanContext is invalid, and all others are valid, even if they have neither trace nor span ID. But this is discussed in #753 and I think out of scope in this PR.

Reviewer 5: well, I think of "SpanContext" as just being a piece of the Span, so same same. I was suggesting a more radical change, rather than hacking it with some id generation rules.

Reviewer 6: If the decision is `DROP`, we don't need to create a Span at all, right? So why we care about the spanid of the span which is not going to be created?

Author: > we don't need to create a Span at all

We need to return a Span instance from the function. And if we use that span as parent, the sampled flag (unset) must be available for the (parent-based) sampler to decide whether to sample the child span. This is required even if that span is in another process, i.e. propagators must have something to propagate.

Author: E.g. let's say we do `A = startSpan()` and A is dropped. Now we create span `B = startSpan(parent=currentContextWith(A))`. If A has no span ID, then this is equivalent to just `B = startSpan()`. However, this would be wrong, since an unsampled parent often causes the child to short-circuit sampling logic and go straight to DROP.

Similarly, if we do `inject(A, httpRequest); httpRequest.execute()` and then in another process `AExtracted = extract(httpRequest); C = startSpan(parent=AExtracted)` we want to know that AExtracted is a unsampled, existing span.

Reviewer 6: Thanks @Oberon00 for clarifying!

Author: @jkwatson I don't know what I could do in this PR to resolve your concerns. Should this discussion maybe be a separate issue?

Reviewer 7: @Oberon00 Is there a downside to using an invalid span ID? I think this isn't really the same case as an extracted non-recording span since that exists for a very short amount of time before creating a real span. In this case, log messages in two different processes will have the same span ID added to them, IIUC, and it does seem like it could cause problems for correlation.

Author: The downside is that a span with invalid ID is indistinguishable from no span, #753

Reviewer 7: Ah - yeah how to handle trace state for this is interesting. I agree it's probably a good idea to decouple OTel from w3c by allowing it to be propagated if it exists, similar to what we do for sampling decision (why not). I think it does have to be addressed separately from the span ID though.

Author: The main topic of #753 is the trace state (which is unrelated to this PR here), but the problem is still that if span ID = 0 then isValid = false, so for example in Java, it would not be injected, and the sampler would ignore it.

Author: > Currently, we consider a SpanContext invalid if the trace or span ID is invalid (i.e. it is invalid unless both are valid), and this is exposed with an API function. As a result, for example in Java, an invalid span context is not inherited by child spans (a fresh context with new span and trace ID is generated, discarding any TraceState or TraceFlags).

Author: I do not want to change the definition of isValid in this PR and I think relying on an invalid span being propagated is bad. Also, even if we did this, we also want to have sampling working with the W3C propagator (with sampled flag=false being propagated), so even if we relax isValid, I still think we need a valid span ID.

Reviewer 7: Ah also I think generating a span ID that's not exported is fine - but similar to @jkwatson's concern sharing the ID (without native support for it, don't forget we already said no to the Zipkin model :) ) is concerning to me from a correlation perspective.

Author: > sharing the ID (without native support for it, don't forget we already said no to the Zipkin model :) ) is concerning to me from a correlation perspective.

I would argue that:

1. Nothing is shared, as a DROP-ed span is not really a span, only a wrapped SpanContext. The span returned by startSpan after a DROP decision where a parent ID is reused will typically be the same spancontext that the parent had wrapped in a SpanContext (for the typical case where we already get an unsampled span when extracting, it can be the same Span instance as in the parent Context even). Backends will never see this span.
2. There is no "native support" for *not* sharing the span ID. The trace will break if a child span becomes sampled again. This is also concerning to me.

For these reasons, I think this is not comparable to the Zipkin model, where information for the same span ID is exported from different places/processes.

Author: So in 2. I mentioned the concrete problem we have with generating a new trace ID. I don't know of any real problem that could result from re-using the parent span ID for non-recording spans (unexported but recording spans might cause trouble, but we are talking about really non-recording spans). If you know of a real problem that might result from this PR, please tell me.

From my POV, this change should be invisible to backends, unless a child-span becomes sampled again, in which case it will be a direct child of the nearest previously sampled parent for the backend, instead of having a missing parent.

Reviewer 5: My concern is not for backends, but for local developer experience. If someone is trying to debug their instrumentation and sees that they are getting duplicate span ids, they're going to scream bloody murder, if they haven't very carefully read the spec and/or documentation.

Author: > If someone is trying to debug their instrumentation and sees that they are getting duplicate span ids, they're going to scream bloody murder, if they haven't very carefully read the spec and/or documentation.

I don't know in which scenario you would usually stumble upon this. I do agree that it may make it a bit harder to debug instrumentations, but I think it might make it easier to debug the monitored systems (because of not breaking the trace).
I fear debugging a dropped span may be difficult anyway, because you need to know about sampling first and if someone sends you something with a `0` sampled flag, you might start scratching your head why your instrumentation is not working for *some* requests.

Reviewer 5: I literally thinking of someone stepping through their code in a debugger. If they don't know the spec well (and most won't!), and they see a newly created span has the same ID as the previous one, there will be an issue logged about it.

Author: They should also notice the strange type of the Span though. And isn't it better to make a dropped span stand out here?

Reviewer 5: Sure, by not having the same span id as another span!  

I'm not going to block this change, but I do think it's a bad idea.

Author: OK, so you would prefer to always generate a new span + trace ID, i.e. treat DROP the same as all other decisions? I think this *is* the other sensible option, although I would prefer the one in this PR. In the end, we need a decision here and I feel the current approvals on this PR are insufficient as this is an important topic.

Reviewer 4: My vote is to reuse span ID, that's the only way I can see to keep the trace structure sensible when some spots in the graph decide not to record - if they don't record, they effective should look like a simple relaying proxy, which _does not_ change span ID.

Reviewer 4: This issue was discussed at the TC meeting today. The TC recommendation is to generate a new span ID even for non-recording spans.

Motivation: @SergeyKanzhelev explained that there are other telemetry use cases that rely on a unique span ID, like log correlations. 

Generating new ID for non-recording span and using that ID as parent ID for subsequent spans will result in broken traces. However, this can be alleviated by extending the logic in the SDK to capture the "last sampled span ID" and using it as a parent.

Author: > However, this can be alleviated by extending the logic in the SDK to capture the "last sampled span ID" and using it as a parent.

How is that logic going to work across the W3C propagator?

Reviewer 8: the idea is that instrumented library needs a reliable unique ID of the operation to associate with logs, exceptions or even return to user as to get support help. Is there is a spanID reuse - library will need to generate an alternative ID for this purposes.

Also platforms may now know all the time if there is a span consumer if has some tricky implementation optimized for hot path execution (like moving all sampling to the end).

The ida is that we will GA with the regenerated spanID and discuss how to improve on it using things like propagating spanID via tracestate. Since we do not have specific design, regeneration looks like a safer choice that will not break SDK going forward and is easier to explain.

Reviewer 8: I planned to write more but forgot it's a veterans day with schools closed today

Author: > regeneration looks like a safer choice that will not break SDK going forward

I think changing from any of them to the other will break some uses.

> and is easier to explain

Fair enough.

> the idea is that instrumented library needs a reliable unique ID of the operation 

If you have an ID that is not associated with any trace on the tracing backend, will that be more helpful than the "inexact" ID of the parent operation? Note, you will have a valid span ID in any case.

I would say we go GA with this, and discuss how to improve on it using things like adding a `GetUniqueOperationId` function to `Span` that returns the span ID if it was generated for that span or generates a new ID that is cached and propagated to children if it is not. I recon propagating this ID cross-process may be difficult, but the same goes for propagating the ID of the last sampled span, and it seems unlikely that you call GetUniqueOperationId on an extracted span directly (it would not be very unique, since the parent's process would have the same ID if it was sampled at least).

Reviewer 8: > > regeneration looks like a safer choice that will not break SDK going forward
> 
> I think changing from any of them to the other will break some uses.
> 
> > and is easier to explain
> 
> Fair enough.
> 
> > the idea is that instrumented library needs a reliable unique ID of the operation
> 
> If you have an ID that is not associated with any trace on the tracing backend, will that be more helpful than the "inexact" ID of the parent operation? Note, you will have a valid span ID in any case.
> 
> I would say we go GA with this, and discuss how to improve on it using things like adding a `GetUniqueOperationId` function to `Span` that returns the span ID if it was generated for that span or generates a new ID that is cached and propagated to children if it is not. I recon propagating this ID cross-process may be difficult, but the same goes for propagating the ID of the last sampled span, and it seems unlikely that you call GetUniqueOperationId on an extracted span directly (it would not be very unique, since the parent's process would have the same ID if it was sampled at least).

The assumption that there is a trace backend may not be true for everybody. And having trace backend of your caller having a record of a parent span Id is not helpful to solve your problem. So the decision we made at @open-telemetry/technical-committee is to go with span ID regeneration. This creates a nice guarantee of local uniqueness of a spanID that various telemetry signals may take a dependency on.

How strongly do you disagree?

Author: Not very strongly. I do think my proposed solution is the better trade-off but I concede that always generating a new ID is also a reasonable option.

Reviewer 2: @Oberon00 Thanks for the answer. I think we will go with the TC decision for now then - if you have time, would you mind updating this PR to match the proposal? Else, I can help by creating a spin-off of this PR. Let me know ;)

Author: I think it would be best to create a spin-off PR. That way we clearly document that this option was rejected.
